torch/fx/proxy.py

```diff
@@ -259,18 +260,29 @@ def __len__(self):
                            "this call to be recorded, please call torch.fx.wrap('len') at "
                            "module scope")
 
-    def __torch_function__(self, orig_method, types, args=None, kwargs=None):
+    @classmethod
+    def __torch_function__(cls, orig_method, types, args=None, kwargs=None):
         args = args if args else ()
         kwargs = kwargs if kwargs else {}
 
+        # Find an instance of this class in the arguments to get a tracer
+        args_of_this_cls = []
+        for a in args:
+            if isinstance(a, cls):
+                args_of_this_cls.append(a)
+            elif isinstance(a, collections.Sequence):
+                args_of_this_cls.extend(el for el in a if isinstance(el, cls))
+        assert len(args_of_this_cls) > 0
+        tracer = args_of_this_cls[0].tracer
```

Comment:
I recommend instead using `map_aggregate` to recursively traverse `args` and `kwargs`:

https://github.com/pytorch/pytorch/blob/57eda692192ed4bcd291f09bf5b2e409c3031db6/torch/fx/node.py#L584

Something like:

```
tracers : Dict[torch.fx.Tracer, None] = []
def find_tracer(a):
  if isinstance(a, cls):
    proxy_inputs.setdefault(a.tracer)
torch.fx.node.map_aggregate(args, find_tracer)
torch.fx.node.map_aggregate(kwargs, find_tracer)

if len(tracers) > 1:
  raise RuntimeError(f'Found multiple different tracers {list(tracers.keys)} while trying to trace operations {orig_method}')
tracer = next(iter(tracers.keys()))
```

Multiple tracers would indeed be an error and we should just throw as in the example code

Reply:
Thanks for the pointer to map_aggregate, that makes it much cleaner.
Updated and fixed lint.